### PR TITLE
AnimExpression: boolean expression parser and evaluator

### PR DESCRIPTION
**Pull request by hyperlogic**
The AnimExpression class is intended to be within animation graphs to evaluate state machine transition expressions.  It's not hooked up to the anim system yet, but is a single unit tested class in the animation library.

It supports identifiers from AnimVariantMap and boolean operators in expressions: '&&', '||' and '!', including parenthesis.  The tokenizer and evaluator also has support for floating point numbers and more complex conditional expressions, but these are not yet supported by the parser.

Unit tests are provided.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: ab85e2967a9bca036abac6b49c5199fba4d4e54c available at: https://gist.github.com/324f4128c3162021ebb1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/High+Fidelity-PR-6660-win64.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6660.dmg

**Comment by ZappoMan** [comment]
should this be in shared and not "Anim" specific -- but instead more generally available for use?

**Comment by hyperlogic** [comment]
@ZappoMan, yes eventually.  Once it is proven to be robust and more full featured I think it can be moved to shared.  But, currently, it only supports boolean expressions and it has a dependency on the AnimVariantMap.

File: tests/animation/src/AnimTests.cpp [hunk 1]
```diff
@@ -394,3 +394,234 @@ void AnimTests::testAnimPose() {
...
+    TEST_BOOL_EXPR(!true && false || true);
+    TEST_BOOL_EXPR(true && !false || !true);
+    TEST_BOOL_EXPR(true && !false || true);
+    TEST_BOOL_EXPR(true && false || !true);
+    TEST_BOOL_EXPR(true && false || true);
+
+    TEST_BOOL_EXPR(!(true && f) || !t);
+    TEST_BOOL_EXPR(!!!(t) && (!!f || true));
+    TEST_BOOL_EXPR(!(true && f) && true);
+}
```
Comment (AndrewMeadows) [inline tests/animation/src/AnimTests.cpp]: Love @hyperlogic for tokenizer unit tests.

**Pull request merged**